lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AvroPageDataReader.java

```diff
@@ -0,0 +1,1065 @@
+import static com.google.common.base.Verify.verify;
+import static io.trino.hive.formats.avro.AvroTypeUtils.UNION_FIELD_TAG_TYPE;
+import static io.trino.hive.formats.avro.AvroTypeUtils.isSimpleNullableUnion;
+import static io.trino.hive.formats.avro.AvroTypeUtils.typeFromAvro;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.lang.Float.floatToRawIntBits;
+import static java.util.Objects.requireNonNull;
+
+public class AvroPageDataReader
+        implements DatumReader<Optional<Page>>
+{
+    private final Schema readerSchema;
+    private Schema writerSchema;
+    private final PageBuilder pageBuilder;
+    private RowBlockBuildingDecoder rowBlockBuildingDecoder;
+    private final AvroTypeManager typeManager;
+
+    public AvroPageDataReader(Schema readerSchema, AvroTypeManager typeManager)
+            throws AvroTypeException
+    {
+        try {
+            this.readerSchema = requireNonNull(readerSchema, "readerSchema is null");
+            this.typeManager = requireNonNull(typeManager, "typeManager is null");
+            this.writerSchema = this.readerSchema;
+            Type readerSchemaType = typeFromAvro(this.readerSchema, typeManager);
+            verify(readerSchemaType instanceof RowType, "Can only build pages when top level type is Row");
+            this.pageBuilder = new PageBuilder(readerSchemaType.getTypeParameters());
+            initialize();
+        }
+        catch (org.apache.avro.AvroTypeException runtimeAvroException) {
+            throw new AvroTypeException(runtimeAvroException.getMessage(), runtimeAvroException.getCause());
+        }
+    }
+
+    private void initialize()
+            throws AvroTypeException
+    {
+        verify(readerSchema.getType().equals(Schema.Type.RECORD), "Avro schema for page reader must be record");
+        verify(writerSchema.getType().equals(Schema.Type.RECORD), "File Avro schema for page reader must be record");
+        this.rowBlockBuildingDecoder = new RowBlockBuildingDecoder(writerSchema, readerSchema, typeManager);
+    }
+
+    protected static class AvroTypeExceptionRuntimeWrapper
+            extends RuntimeException
+    {
+        private final AvroTypeException avroTypeException;
+
+        public AvroTypeExceptionRuntimeWrapper(AvroTypeException cause)
+        {
+            super(cause);
+            this.avroTypeException = cause;
+        }
+
+        public AvroTypeException getAvroTypeException()
+        {
+            return avroTypeException;
+        }
+    }
+
+    @Override
+    public void setSchema(Schema schema)
+    {
+        if (schema != null && schema != writerSchema) {
```

Comment:
Does this ever get called with a null?  If it does not happen, add null check.  If it does, add a comment on why that is allowed... something like `// a null schema means the file does not have a recorded schema, so we keep the original requested schema in the constructor`